src/server/rpc/procedures/utils/api-consumer.js

```diff
@@ -0,0 +1,259 @@
+                this._logger.error('invalid id / response',res.headers);
+                imgResponse.isImage = false;
+                this.response.send('null');
+            }
+        });
+        return imgResponse;
+    }
+
+    // private
+    _sendNext() {
+        var msgs = remainingMsgs[this.socket.uuid];
+        if (msgs && msgs.length) {
+            var msg = msgs.shift();
+
+            // while (msgs.length && msg.dstId !== this.socket.uuid) {
+            //     msg = msgs.shift();
+            // }
+
+            // check that the socket is still at the role receiving the messages
+            if (msg && msg.dstId === this.socket.roleId) {
+                this._logger.trace('sending msg to', this.socket.uuid, this.socket.roleId);
+                this.socket.send(msg);
+            }
+
+            if (msgs.length) {
+                setTimeout(this._sendNext.bind(this), MSG_SENDING_DELAY);
+            } else {
+                delete remainingMsgs[this.socket.uuid];
+            }
+        } else {
+            delete remainingMsgs[this.socket.uuid];
+        }
+    }
+
+    /**
+     * processes and queries json object or strings
+     * @param  {json/string} json  [description]
+     * @param  {string} query query string written in standard js notation: '.attr.attr[i]'
+     * @return {json}       returns the value found withing the input json
+     */
+    _queryJson(json, query){
+      // assuming that there is no digit in attribute names
+        if (typeof(json) === 'string') {
+            json = JSON.parse(json);
+        }
+        if(!query){
+            return json;
+        }
+        let queryComponents = [],
+            res = json;
+        query.split('.').forEach(item => {
+            let searchRes = /\d+/g.exec(item);
+            if (searchRes) {
+                queryComponents.push(item.substring(0,searchRes.index -1));
+                queryComponents.push(searchRes[0]);
+            }else {
+                queryComponents.push(item);
+            }
+        });
+        queryComponents.shift(); // remove the first item which is always empty=
+        queryComponents.forEach(q=>{
+            res = res[q] ? res[q] : null;
+        });
+        // if (typeof(res) === 'object') {
+        //     res = JSON.stringify(res);
+        // }
+        if (res === null) this._logger.trace(query, 'query, returned no result');
+        return res;
+    }
+
+
+    /**
+     * request a full response sending back a data structure.
+     * @param  {string} queryOptions
+     * @param  {Function} parseFn should convert a valid response from the api into an array of simple (tuple - keyvalue) Json Objects for  messages
+     * @return {Promise}
+     */
+    _sendStruct(queryOptions,parserFn){
+        return this._requestData(queryOptions)
+            .then(res => {
+                this._logger.trace('got response back',res);
+                let parsedRes =  parserFn(res);
```

Comment:
Do you think I should a try catch around this parser ? then when the user is writing the parserFn can throw errors when those happen.